Sources/GRPCHTTP2Core/GRPCStreamStateMachine.swift

```diff
@@ -0,0 +1,222 @@
+    var compressor: Zlib.Compressor?
+    
+    init(
+      maximumPayloadSize: Int,
+      compressor: Zlib.Compressor?,
+      decompressor: Zlib.Decompressor?
+    ) {
+      self.framer = GRPCMessageFramer()
+      let messageDeframer = GRPCMessageDeframer(
+        maximumPayloadSize: maximumPayloadSize,
+        decompressor: decompressor
+      )
+      self.deframer = NIOSingleStepByteToMessageProcessor(messageDeframer)
+      self.compressor = compressor
+    }
+  }
+  
+  init(configuration: Configuration) {
+    self.state = .clientIdleServerIdle(configuration)
+  }
+  
+  mutating func send(metadata: Metadata) throws {
+    // Client only sends metadata when opening.
+    try self.clientSentMetadata()
+  }
+  
+  mutating func send(message: [UInt8]) {
+    // Client sends message.
+    switch self.state {
+    case .clientIdleServerIdle(let configuration):
+      preconditionFailure("Client not yet open")
+    case .clientOpenServerIdle(var clientOpenState):
+      clientOpenState.framer.append(message)
+      self.state = .clientOpenServerClosed(clientOpenState)
+    case .clientOpenServerOpen(var clientOpenState):
+      clientOpenState.framer.append(message)
+      self.state = .clientOpenServerClosed(clientOpenState)
+    case .clientOpenServerClosed:
+      // Nothing to do: no point in sending a message if the server's closed.
+      ()
+    case .clientClosedServerIdle, .clientClosedServerOpen, .clientClosedServerClosed:
+      preconditionFailure("Client is closed, cannot send a message")
+    }
+  }
+  
+  mutating func send(status: String, trailingMetadata: Metadata) throws {
+    // Only server does this.
+  }
+  
+  /// Returns the client's next request to the server.
+  /// - Returns: The request to be made to the server.
+  mutating func nextRequest() throws -> ByteBuffer? {
+    switch self.state {
+    case .clientIdleServerIdle(let configuration):
+      preconditionFailure("Client is not open yet.")
+    case .clientOpenServerIdle(var clientOpenState):
+      let request = try clientOpenState.framer.next(compressor: clientOpenState.compressor)
+      self.state = .clientOpenServerIdle(clientOpenState)
+      return request
+    case .clientOpenServerOpen(var clientOpenState):
+      let request = try clientOpenState.framer.next(compressor: clientOpenState.compressor)
+      self.state = .clientOpenServerOpen(clientOpenState)
+      return request
+    case .clientOpenServerClosed:
+      // Nothing to do: no point in sending request if server is closed.
+      return nil
+    case .clientClosedServerIdle, .clientClosedServerOpen, .clientClosedServerClosed:
+      preconditionFailure("Can't send request if client is closed")
+    }
+  }
+  
+  mutating func receive(metadata: Metadata, endStream: Bool) {
+    // This is metadata received by the client from the server.
+    // It can be initial, which confirms that the server is now open;
+    // or an END_STREAM trailer, meaning the response is over.
```

Comment:
The server might send trailers without end stream set and then follow it with an empty data frame with end stream set, so we also need to be wary of that. One way to "know" they are meant as trailers is the presence of 'grpc-status' . This check could be done by the handler though, in essence "faking" the end stream, I think that'd be fine.